### ISSUE TITLE
In our experiment, origin lightgbm predict result is not equal treelite predict result

### ISSUE THREAD

**Issue by henriezhang** [dmlc/treelite#77]
Dependencies:
lightgbm==2.2.2
treelite==0.32

LightGBM was trained with the following parameters:
boosting_type = gbdt
objective = binary
metric = binary_logloss,auc
metric_freq = 1
is_training_metric = true
max_bin = 255
categorical_feature=1,2,3,5,6,8,299
scale_pos_weight = 2.0
num_trees = 500
num_leaves = 255
max_depth = 20
learning_rate = 0.1
min_gain_to_split = 0.00000001
feature_fraction = 0.7
bagging_freq = 5
bagging_fraction = 0.8
min_data_in_leaf = 10
min_sum_hessian_in_leaf = 1e-3
is_enable_sparse = true
use_two_round_loading = true

Save the model and predict sum case:
the result is  as follow:
 0.077086254685271596
 0.044846837005837775
 0.66469752969639961

compile the model and predict the samples:
model = treelite.Model.load('./a.mdl', model_format='lightgbm')
model.export_lib(toolchain='gcc', libpath='./a.so', verbose=True)

batch = treelite.runtime.Batch.from_npy2d(matrix)
predictor = treelite.runtime.Predictor('./a.so', verbose=True)
out_pred = predictor.predict(batch, pred_margin=True)
the reuslt is different from origin model predict:
0.10160005 
0.05897592 
0.6094563

**Comment by henriezhang** [comment]
when set the num_trees=1 the result is similar:
origin model result:
![image](https://user-images.githubusercontent.com/6031726/53138169-56564c80-35c0-11e9-8141-bbe78b83fa54.png)

treelite retule:
![image](https://user-images.githubusercontent.com/6031726/53138186-6706c280-35c0-11e9-901f-cf9f3d1cb261.png)


when set num_trees=2 the result is begin different:
origin model result:
![image](https://user-images.githubusercontent.com/6031726/53138224-8aca0880-35c0-11e9-8450-66777c81a318.png)
treelite result:
![image](https://user-images.githubusercontent.com/6031726/53138238-97e6f780-35c0-11e9-9c70-e81227b717cc.png)

**Comment by hcho3** [comment]
@henriezhang Is it possible to post your model?

**Comment by henriezhang** [comment]
> @henriezhang Is it possible to post your model?
model is more than 2G, how post to you?

**Comment by henriezhang** [comment]
my model is more than 2G, how post to you?

**Comment by hcho3** [comment]
@henriezhang Dropbox or Google Drive link will work. I’ll make sure the bug is fixed.

**Comment by henriezhang** [comment]
please give me your email? I send you a simple model have the same problem

**Comment by hcho3** [comment]
chohyu01@cs.washington.edu

**Comment by henriezhang** [comment]
@hcho3 I have sanded you the model to your email.

**Comment by hcho3** [comment]
@henriezhang #81 should fix the bug. Thanks so much for reporting!

**Comment by henriezhang** [comment]
@hcho3 well done!

**Comment by henuxhj** [comment]
On the same experiment parameters（num_tree=500, depth=20, leaves=255 ）, treelite cost more memory than before， It generate more source code than be before.  And treelite write source code on disk after generate total source code.  It may cause OOM . 
Is there any way to solve this problem?
Thanks

**Comment by hcho3** [comment]
@henuxhj This is because your model has high cardinality categorical features, and Treelite used to truncate high categorical values. So the larger code you get is the correct code (in the sense of giving correct prediction).

Please set `parallel_comp` option when compiling, to reduce memory consumption. This will break the 500 tree models into smaller pieces.

```python
model.export_lib(toolchain='gcc', libpath='./a.so',
                verbose=True, params={'parallel_comp':500})

**Comment by henuxhj** [comment]
thank you so much， I'll try.
